oci/pack.go

```diff
@@ -443,11 +459,18 @@ unpack:
 
 		// Set the initrd if available
 		initrdPath := filepath.Join(popts.Workdir(), WellKnownInitrdPath)
-		if f, err := os.Stat(initrdPath); err == nil && f.Size() > 0 {
+		if f, err := os.Stat(initrdPath); err == nil && !f.IsDir() && f.Size() > 0 {
 			ocipack.initrd, err = initrd.NewFromFile(popts.Workdir(), initrdPath)
 			if err != nil {
 				return err
 			}
+		} else if err == nil && f.IsDir() {
+			ocipack.initrd, err = initrd.NewFromMapping(popts.Workdir(),
+				filepath.Join(filepath.Dir(initrdPath), "initramfs.cpio"),
+				fmt.Sprintf("%s:/", initrdPath))
+			if err != nil {
+				return err
+			}
```

Comment:
A few thoughts on this part specifically. The general implementation of this PR allows for serializing a filesystem into a single layer (which in itself outside of the context of using it as initramfs is very useful), and here we dynamically serialize it into a initramfs when provided with a directory AFTER unpacking it. I imagined when a user serialized a filesystem as a "normal" OCI layer (tarball of a directory) that this may not necessarily be initramfs (because it can be both initramfs OR it can act like a normal filesystem and be extracted somewhere. even one day multiple layers can be "overlayed" like how the normal OCI spec dictates). I think the tarball work is useful but should be opened up as a separate PR.

To simplify this, i would move this `initrd.NewFromMapping` to much earlier, where `NewLayerFromDirectory` is actually called. This way we pack the initramfs at build time. This has the added benefit of speeding up general instantiation because we don't have to do any "post processing" before instantiating the unikernel (untarring then turning into a CPIO).

Reply:
Applied your changes.

Did not open an new PR because It's more or less the same thing 🪄